### PR TITLE
Generic hashing for the new closure representation

### PR DESCRIPTION
**Pull request by xavierleroy**
This PR revises the generic hash functions to take advantage of the new closure representation introduced in #9619: code pointers are directly mixed into the hash rather than having to be detected using `Is_in_value_area`.

As a consequence, the `Is_in_value_area` test can be removed altogether in no-naked-pointers mode, which should make hashing faster.

Currently the new code for closures is activated only in no-naked-pointers mode, even though it is sound in naked-pointers mode too.

Closes: #2168

### CODE REVIEW

**Comment by stedolan** [comment]
Since closures are supported both in marshalling and in hashing, it would be nice if marshalling and unmarshalling a closure did not change its hash.

This would have been painful to achieve before, but I think it can now be done easily enough by hashing `codeptr - cf->code_start` instead of `codeptr`, after looking up `cf` using the approach of #9649?

**Comment by xavierleroy** [comment]
> hashing codeptr - cf->code_start instead of codeptr

This can be done, and with #9654 it might even be not too slow :-)

For additional fun, we could even mix in (some bits of) the digest for the code fragment.  But this may be going too far.

One caveat, though: I'd like to have the same hash values in naked-pointers mode and in no-naked-pointers mode.  So, this means activating the new hashing schema for closures in both modes, and not just in no-naked-pointers mode, as done conservatively in this PR.

**Comment by stedolan** [comment]
> One caveat, though: I'd like to have the same hash values in naked-pointers mode and in no-naked-pointers mode. So, this means activating the new hashing schema for closures in both modes, and not just in no-naked-pointers mode, as done conservatively in this PR.

That would indeed be nice, but even having deterministic hash values in either of these modes would be an improvement. Currently, and even with this PR, closures do not have the same hash values between two runs of the same binary. (At least on PIE systems with ASLR, or in the presence of dynlink)

**Comment by xavierleroy** [comment]
OK, I'll implement the "relative code pointer" approach.  Since #9654 changes the handling of code fragments, I'd rather wait until #9654 is reviewed and merged.

**Comment by xavierleroy** [comment]
Rebased on trunk and implemented @stedolan's "relative code pointer" approach.  This is ready for review.

**Comment by xavierleroy** [comment]
This needs to be reviewed.  @stedolan ? @damiendoligez ? Please ?

**Review by stedolan — APPROVED** [review]

**Comment by stedolan** [comment]
Did this get merged?

**Comment by xavierleroy** [comment]
Oups, I made a mistake...

**Comment by xavierleroy** [comment]
I can't resurrect this PR, so I made a new one at #9689.  Apologies for my clumsiness.

**Pull request closed without merging**
